File: python/features/steps/issue_creation_steps.py
"""Behave steps for issue creation."""

from __future__ import annotations

from behave import then, when
from pathlib import Path
from types import SimpleNamespace

from features.steps.shared import (
    capture_issue_identifier,
    load_project_directory,
    read_issue_file,
    run_cli,
)
from kanbus.issue_creation import IssueCreationError, create_issue


@when('I run "kanbus create Implement OAuth2 flow"')
def when_run_create_default(context: object) -> None:
    run_cli(context, "kanbus create Implement OAuth2 flow")


@when('I run "kanbus create implement oauth2 flow"')
def when_run_create_duplicate_title(context: object) -> None:
    run_cli(context, "kanbus create implement oauth2 flow")


@when(
    'I run "kanbus create Fix login bug --type bug --priority 1 --assignee dev@example.com --parent kanbus-epic01 --label auth --label urgent --description \\"Bug in login\\""'
)
def when_run_create_full(context: object) -> None:
    run_cli(
        context,
        "kanbus create Fix login bug --type bug --priority 1 --assignee dev@example.com "
        '--parent kanbus-epic01 --label auth --label urgent --description "Bug in login"',
    )


@when('I run "kanbus create Bad Issue --type nonexistent"')
def when_run_create_invalid_type(context: object) -> None:
    run_cli(context, "kanbus create Bad Issue --type nonexistent")


@when('I run "kanbus create Orphan --parent kanbus-nonexistent"')
def when_run_create_missing_parent(context: object) -> None:
    run_cli(context, "kanbus create Orphan --parent kanbus-nonexistent")


@when('I run "kanbus create"')
def when_run_create_without_title(context: object) -> None:
    run_cli(context, "kanbus create")


@when('I run "kanbus create Bad Priority --priority 99"')
def when_run_create_invalid_priority(context: object) -> None:
    run_cli(context, "kanbus create Bad Priority --priority 99")


@when(
    'I run "kanbus create Bad Parent --type epic --parent kanbus-epic01 --no-validate"'
)
def when_run_create_invalid_parent_no_validate(context: object) -> None:
    run_cli(
        context,
        "kanbus create Bad Parent --type epic --parent kanbus-epic01 --no-validate",
    )


@when('I run "kanbus create Child Task --type {issue_type} --parent kanbus-parent"')
def when_run_create_child_task_with_parent(context: object, issue_type: str) -> None:
    run_cli(
        context,
        f"kanbus create Child Task --type {issue_type} --parent kanbus-parent",
    )


@when('I run "kanbus create Child --type {issue_type} --parent kanbus-bug01"')
def when_run_create_child_with_bug_parent(context: object, issue_type: str) -> None:
    run_cli(
        context,
        f"kanbus create Child --type {issue_type} --parent kanbus-bug01",
    )


@when('I run "kanbus create Standalone Task --type task"')
def when_run_create_standalone_task(context: object) -> None:
    run_cli(context, "kanbus create Standalone Task --type task")


@when('I run "kanbus create Snapshot issue"')
def when_run_create_snapshot_issue(context: object) -> None:
    run_cli(context, "kanbus create Snapshot issue")


@when('I create an issue directly with title "Implement OAuth2 flow"')
def when_create_issue_directly(context: object) -> None:
    working_directory = getattr(context, "working_directory", None)
    if working_directory is None:
        raise RuntimeError("working directory not set")
    root = Path(working_directory)
    try:
        create_issue(
            root=root,
            title="Implement OAuth2 flow",
            issue_type=None,
            priority=None,
            assignee=None,
            parent=None,
            labels=[],
            description="",
            local=False,
        )
    except IssueCreationError as error:
        context.result = SimpleNamespace(
            exit_code=1,
            stdout="",
            stderr=str(error),
            output=str(error),
        )
        return
    context.result = SimpleNamespace(exit_code=0, stdout="", stderr="", output="")


@then("the command should succeed")
def then_command_succeeds(context: object) -> None:
    if context.result.exit_code != 0:
        print(f"Command failed with exit code {context.result.exit_code}")
        print(f"STDOUT: {context.result.stdout}")
        print(f"STDERR: {context.result.stderr}")
    assert context.result.exit_code == 0


@then("stdout should contain a valid issue ID")
def then_stdout_contains_issue_id(context: object) -> None:
    capture_issue_identifier(context)


@then("an issue file should be created in the issues directory")
def then_issue_file_created(context: object) -> None:
    project_dir = load_project_directory(context)
    issues = list((project_dir / "issues").glob("*.json"))
    assert len(issues) == 1


@then("the issues directory should contain {issue_count:d} issue file")
def then_issues_directory_contains_count(context: object, issue_count: int) -> None:
    project_dir = load_project_directory(context)
    issues = list((project_dir / "issues").glob("*.json"))
    assert len(issues) == issue_count


@then('the created issue should have title "Implement OAuth2 flow"')
def then_created_issue_title(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.title == "Implement OAuth2 flow"


@then('the created issue should have type "task"')
def then_created_issue_type(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.issue_type == "task"


@then('the created issue should have status "open"')
def then_created_issue_status(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.status == "open"


@then("the created issue should have priority 2")
def then_created_issue_priority(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.priority == 2


@then("the created issue should have an empty labels list")
def then_created_issue_labels_empty(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.labels == []


@then("the created issue should have an empty dependencies list")
def then_created_issue_dependencies_empty(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.dependencies == []


@then("the created issue should have a created_at timestamp")
def then_created_issue_created_at(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.created_at is not None


@then("the created issue should have an updated_at timestamp")
def then_created_issue_updated_at(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.updated_at is not None


@then('the created issue should have type "bug"')
def then_created_issue_type_bug(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.issue_type == "bug"


@then("the created issue should have priority 1")
def then_created_issue_priority_one(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.priority == 1


@then('the created issue should have assignee "dev@example.com"')
def then_created_issue_assignee(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.assignee == "dev@example.com"


@then('the created issue should have parent "kanbus-epic01"')
def then_created_issue_parent(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.parent == "kanbus-epic01"


@then('the created issue should have labels "auth, urgent"')
def then_created_issue_labels(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.labels == ["auth", "urgent"]


@then('the created issue should have description "Bug in login"')
def then_created_issue_description(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.description == "Bug in login"


@then("the created issue should have no parent")
def then_created_issue_no_parent(context: object) -> None:
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.parent is None

File: rust/features/steps/issue_creation_steps.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

use cucumber::{given, then, when};
use regex::Regex;
use serde_json::Value;
use std::env;
use tempfile::TempDir;

use kanbus::cli::run_from_args_with_output;
use kanbus::file_io::load_project_directory;
use kanbus::issue_creation::{create_issue, IssueCreationRequest};

use crate::step_definitions::initialization_steps::KanbusWorld;

fn run_cli(world: &mut KanbusWorld, command: &str) {
    let args = shell_words::split(command).expect("parse command");
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");

    match run_from_args_with_output(args, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[when("I create an issue directly with title \"Implement OAuth2 flow\"")]
fn when_create_issue_directly(world: &mut KanbusWorld) {
    let root = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let request = IssueCreationRequest {
        root: root.clone(),
        title: "Implement OAuth2 flow".to_string(),
        issue_type: None,
        priority: None,
        assignee: None,
        parent: None,
        labels: Vec::new(),
        description: None,
        local: false,
        validate: true,
    };
    match create_issue(&request) {
        Ok(_) => {
            world.exit_code = Some(0);
            world.stdout = Some(String::new());
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

fn load_project_dir(world: &KanbusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("cwd");
    load_project_directory(cwd).expect("project dir")
}

fn capture_issue_identifier(world: &mut KanbusWorld) -> String {
    let stdout = world.stdout.as_ref().expect("stdout");
    let ansi_regex = Regex::new(r"\x1b\[[0-9;]*m").expect("regex");
    let clean_stdout = ansi_regex.replace_all(stdout, "");
    let full_regex = Regex::new(
        r"([A-Za-z0-9]+-[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})",
    )
    .expect("regex");
    if let Some(capture) = full_regex
        .captures(&clean_stdout)
        .and_then(|matches| matches.get(1))
        .map(|match_value| match_value.as_str().to_string())
    {
        world.stdout = Some(stdout.to_string());
        return capture;
    }

    let labeled_regex = Regex::new(r"(?i)\bID:\s*([A-Za-z0-9.-]+)").expect("regex");
    let abbreviated = labeled_regex
        .captures(&clean_stdout)
        .and_then(|matches| matches.get(1))
        .map(|match_value| match_value.as_str().to_string())
        .unwrap_or_else(|| {
            let fallback_regex = Regex::new(r"\b([A-Za-z0-9]{6}(?:\.[0-9]+)?)\b").expect("regex");
            fallback_regex
                .captures(&clean_stdout)
                .and_then(|matches| matches.get(1))
                .map(|match_value| match_value.as_str().to_string())
                .expect("issue id not found")
        });
    let (abbrev_base, abbrev_suffix) = parse_abbreviation(&abbreviated);
    let project_dir = load_project_dir(world);
    let issues_dir = project_dir.join("issues");
    let entries = fs::read_dir(issues_dir).expect("read issues dir");
    for entry in entries {
        let path = entry.expect("issue entry").path();
        if path.extension().and_then(|ext| ext.to_str()) != Some("json") {
            continue;
        }
        let identifier = path
            .file_stem()
            .and_then(|stem| stem.to_str())
            .unwrap_or_default()
            .to_string();
        if matches_abbreviation(&identifier, &abbrev_base, abbrev_suffix.as_deref()) {
            world.stdout = Some(stdout.to_string());
            return identifier;
        }
    }
    panic!("issue id not found");
}

fn parse_abbreviation(value: &str) -> (String, Option<String>) {
    let remainder = if let Some((_, rest)) = value.split_once('-') {
        rest
    } else {
        value
    };
    if let Some((base, suffix)) = remainder.split_once('.') {
        (base.to_lowercase(), Some(suffix.to_lowercase()))
    } else {
        (remainder.to_lowercase(), None)
    }
}

fn matches_abbreviation(identifier: &str, base: &str, suffix: Option<&str>) -> bool {
    let remainder = if let Some((_, rest)) = identifier.split_once('-') {
        rest
    } else {
        identifier
    };
    let (id_base, id_suffix) = if let Some((head, tail)) = remainder.split_once('.') {
        (head, Some(tail))
    } else {
        (remainder, None)
    };
    let normalized = id_base.replace('-', "").to_lowercase();
    if !normalized.starts_with(base) {
        return false;
    }
    match (suffix, id_suffix) {
        (None, _) => true,
        (Some(expected), Some(actual)) => actual.to_lowercase() == expected.to_lowercase(),
        _ => false,
    }
}

fn load_issue_json(project_dir: &PathBuf, identifier: &str) -> Value {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    let contents = fs::read_to_string(issue_path).expect("read issue");
    serde_json::from_str(&contents).expect("parse issue")
}

#[given("a Kanbus project with default configuration")]
fn given_kanbus_project(world: &mut KanbusWorld) {
    env::set_var("KANBUS_NO_DAEMON", "1");
    let temp_dir = TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("repo");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    world.working_directory = Some(repo_path);
    world.temp_dir = Some(temp_dir);
    run_cli(world, "kanbus init");
    assert_eq!(world.exit_code, Some(0));
}

#[when("I run \"kanbus create Implement OAuth2 flow\"")]
fn when_run_create_default(world: &mut KanbusWorld) {
    run_cli(world, "kanbus create Implement OAuth2 flow");
}

#[when("I run \"kanbus create implement oauth2 flow\"")]
fn when_run_create_duplicate_title(world: &mut KanbusWorld) {
    run_cli(world, "kanbus create implement oauth2 flow");
}

#[when("I run \"kanbus create Fix login bug --type bug --priority 1 --assignee dev@example.com --parent kanbus-epic01 --label auth --label urgent --description \\\"Bug in login\\\"\"")]
fn when_run_create_full(world: &mut KanbusWorld) {
    run_cli(world, "kanbus create Fix login bug --type bug --priority 1 --assignee dev@example.com --parent kanbus-epic01 --label auth --label urgent --description \"Bug in login\"");
}

#[when("I run \"kanbus create Bad Issue --type nonexistent\"")]
fn when_run_create_invalid_type(world: &mut KanbusWorld) {
    run_cli(world, "kanbus create Bad Issue --type nonexistent");
}

#[when("I run \"kanbus create Bad Parent --type epic --parent kanbus-epic01 --no-validate\"")]
fn when_run_create_invalid_parent_no_validate(world: &mut KanbusWorld) {
    run_cli(
        world,
        "kanbus create Bad Parent --type epic --parent kanbus-epic01 --no-validate",
    );
}

#[when("I run \"kanbus create Orphan --parent kanbus-nonexistent\"")]
fn when_run_create_missing_parent(world: &mut KanbusWorld) {
    run_cli(world, "kanbus create Orphan --parent kanbus-nonexistent");
}

#[when("I run \"kanbus create\"")]
fn when_run_create_without_title(world: &mut KanbusWorld) {
    run_cli(world, "kanbus create");
}

#[when("I run \"kanbus create Bad Priority --priority 99\"")]
fn when_run_create_invalid_priority(world: &mut KanbusWorld) {
    run_cli(world, "kanbus create Bad Priority --priority 99");
}

#[when(expr = "I run \"kanbus create Child Task --type {word} --parent kanbus-parent\"")]
fn when_run_create_child_task_with_parent(world: &mut KanbusWorld, issue_type: String) {
    run_cli(
        world,
        &format!("kanbus create Child Task --type {issue_type} --parent kanbus-parent"),
    );
}

#[when(expr = "I run \"kanbus create Child --type {word} --parent kanbus-bug01\"")]
fn when_run_create_child_with_bug_parent(world: &mut KanbusWorld, issue_type: String) {
    run_cli(
        world,
        &format!("kanbus create Child --type {issue_type} --parent kanbus-bug01"),
    );
}

#[when("I run \"kanbus create Standalone Task --type task\"")]
fn when_run_create_standalone_task(world: &mut KanbusWorld) {
    run_cli(world, "kanbus create Standalone Task --type task");
}

#[when("I run \"kanbus create Snapshot issue\"")]
fn when_run_create_snapshot_issue(world: &mut KanbusWorld) {
    run_cli(world, "kanbus create Snapshot issue");
}

#[then("the command should succeed")]
fn then_command_succeeds(world: &mut KanbusWorld) {
    assert_eq!(
        world.exit_code,
        Some(0),
        "stderr: {:?}",
        world.stderr.as_deref().unwrap_or("")
    );
}

#[then("stdout should contain a valid issue ID")]
fn then_stdout_contains_issue_id(world: &mut KanbusWorld) {
    let _ = capture_issue_identifier(world);
}

#[then("an issue file should be created in the issues directory")]
fn then_issue_file_created(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issues_dir = project_dir.join("issues");
    let count = fs::read_dir(&issues_dir)
        .expect("read issues dir")
        .filter(|entry| {
            let Ok(item) = entry else {
                return false;
            };
            let path = item.path();
            path.extension()
                .and_then(|ext| ext.to_str())
                .map(|ext| ext == "json")
                .unwrap_or(false)
        })
        .count();
    assert_eq!(count, 1);
}

#[then(expr = "the issues directory should contain {int} issue file")]
fn then_issues_directory_contains_count(world: &mut KanbusWorld, count: i32) {
    let project_dir = load_project_dir(world);
    let issues_dir = project_dir.join("issues");
    let actual = fs::read_dir(&issues_dir)
        .expect("read issues dir")
        .filter(|entry| {
            let Ok(item) = entry else {
                return false;
            };
            let path = item.path();
            path.extension()
                .and_then(|ext| ext.to_str())
                .map(|ext| ext == "json")
                .unwrap_or(false)
        })
        .count();
    assert_eq!(actual as i32, count);
}

#[then("the created issue should have title \"Implement OAuth2 flow\"")]
fn then_created_issue_title(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["title"], "Implement OAuth2 flow");
}

#[then("the created issue should have type \"task\"")]
fn then_created_issue_type_task(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["type"], "task");
}

#[then("the created issue should have status \"open\"")]
fn then_created_issue_status(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["status"], "open");
}

#[then("the created issue should have priority 2")]
fn then_created_issue_priority(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["priority"], 2);
}

#[then("the created issue should have an empty labels list")]
fn then_created_issue_labels_empty(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["labels"].as_array().map(Vec::len), Some(0));
}

#[then("the created issue should have an empty dependencies list")]
fn then_created_issue_dependencies_empty(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["dependencies"].as_array().map(Vec::len), Some(0));
}

#[then("the created issue should have a created_at timestamp")]
fn then_created_issue_created_at(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert!(payload.get("created_at").is_some());
}

#[then("the created issue should have an updated_at timestamp")]
fn then_created_issue_updated_at(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert!(payload.get("updated_at").is_some());
}

#[then("the created issue should have type \"bug\"")]
fn then_created_issue_type_bug(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["type"], "bug");
}

#[then("the created issue should have priority 1")]
fn then_created_issue_priority_one(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["priority"], 1);
}

#[then("the created issue should have assignee \"dev@example.com\"")]
fn then_created_issue_assignee(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["assignee"], "dev@example.com");
}

#[then("the created issue should have parent \"kanbus-epic01\"")]
fn then_created_issue_parent(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["parent"], "kanbus-epic01");
}

#[then("the created issue should have labels \"auth, urgent\"")]
fn then_created_issue_labels(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    let labels = payload["labels"]
        .as_array()
        .expect("labels array")
        .iter()
        .filter_map(|value| value.as_str())
        .collect::<Vec<_>>();
    assert_eq!(labels, vec!["auth", "urgent"]);
}

#[then("the created issue should have description \"Bug in login\"")]
fn then_created_issue_description(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert_eq!(payload["description"], "Bug in login");
}

#[then("the created issue should have no parent")]
fn then_created_issue_no_parent(world: &mut KanbusWorld) {
    let identifier = capture_issue_identifier(world);
    let project_dir = load_project_dir(world);
    let payload = load_issue_json(&project_dir, &identifier);
    assert!(payload["parent"].is_null());
}
